pkg/resources/api_integration.go

```diff
@@ -127,143 +130,186 @@ func APIIntegration() *schema.Resource {
 	}
 
-	if _, ok := d.GetOk("comment"); ok {
-		stmt.SetString("COMMENT", d.Get("comment").(string))
-	}
-
-	// Now, set the API provider
-	if err := setAPIProviderSettings(d, stmt); err != nil {
-		return err
+	apiProvider := d.Get("api_provider").(string)
+	switch apiProvider {
+	case "aws_api_gateway", "aws_private_api_gateway", "aws_gov_api_gateway", "aws_gov_private_api_gateway":
+		roleArn, ok := d.GetOk("api_aws_role_arn")
+		if !ok {
+			return fmt.Errorf("if you use AWS api provider you must specify an api_aws_role_arn")
+		}
+		awsParams := sdk.NewAwsApiParamsRequest(sdk.ApiIntegrationAwsApiProviderType(apiProvider), roleArn.(string))
+		if v, ok := d.GetOk("api_key"); ok {
+			awsParams.WithApiKey(sdk.String(v.(string)))
+		}
+		createRequest.WithAwsApiProviderParams(awsParams)
+	case "azure_api_management":
+		tenantId, ok := d.GetOk("azure_tenant_id")
+		if !ok {
+			return fmt.Errorf("if you use the Azure api provider you must specify an azure_tenant_id")
+		}
+		applicationId, ok := d.GetOk("azure_ad_application_id")
+		if !ok {
+			return fmt.Errorf("if you use the Azure api provider you must specify an azure_ad_application_id")
+		}
+		azureParams := sdk.NewAzureApiParamsRequest(tenantId.(string), applicationId.(string))
+		if v, ok := d.GetOk("api_key"); ok {
+			azureParams.WithApiKey(sdk.String(v.(string)))
+		}
+		createRequest.WithAzureApiProviderParams(azureParams)
+	case "google_api_gateway":
+		audience, ok := d.GetOk("google_audience")
+		if !ok {
+			return fmt.Errorf("if you use GCP api provider you must specify a google_audience")
+		}
+		googleParams := sdk.NewGoogleApiParamsRequest(audience.(string))
+		createRequest.WithGoogleApiProviderParams(googleParams)
+	default:
+		return fmt.Errorf("unexpected provider %v", apiProvider)
 	}
 
-	if err := snowflake.Exec(db, stmt.Statement()); err != nil {
+	err := client.ApiIntegrations.Create(ctx, createRequest)
+	if err != nil {
 		return fmt.Errorf("error creating api integration: %w", err)
 	}
 
-	d.SetId(name)
+	d.SetId(helpers.EncodeSnowflakeID(id))
 
 	return ReadAPIIntegration(d, meta)
 }
 
 // ReadAPIIntegration implements schema.ReadFunc.
 func ReadAPIIntegration(d *schema.ResourceData, meta interface{}) error {
 	db := meta.(*sql.DB)
-	id := d.Id()
-
-	stmt := snowflake.NewAPIIntegrationBuilder(id).Show()
-	row := snowflake.QueryRow(db, stmt)
-
-	// Some properties can come from the SHOW INTEGRATION call
+	ctx := context.Background()
+	client := sdk.NewClientFromDB(db)
+	id := helpers.DecodeSnowflakeID(d.Id()).(sdk.AccountObjectIdentifier)
 
-	s, err := snowflake.ScanAPIIntegration(row)
+	integration, err := client.ApiIntegrations.ShowByID(ctx, id)
 	if err != nil {
-		// If no such resource exists, it is not an error but rather not exist
-		if err.Error() == snowflake.ErrNoRowInRS {
-			d.SetId("")
-			return nil
-		}
-		return fmt.Errorf("could not show api integration: %w", err)
+		log.Printf("[DEBUG] api integration (%s) not found", d.Id())
```

Comment:
one thing we should discuss is how if you don't have required permissions, then the show command will return error same as if the object didn't exist. if customer applies using accountadmin, then later with a less privileged role, then their terraform state file will be completely wiped out. should we check that user has sufficient privileges to do show command before setting resource id to empty? I think its not necessary, but it might be a good improvement later.

Reply:
That's true for all the resources, so I am adding this to SNOW-867235.